Expose atomic bitwise functions (invoked via in-place operator)

j3soon opened the pull request:
Expose `atomic_and`, `atomic_or`, `atomic_xor`, and invoke them through in-place operators `&=`, `|=`, `^=`.

Fixes: https://github.com/NVIDIA/warp/issues/886

<!--
Thank you for contributing to NVIDIA Warp!

See the contribution guide: https://nvidia.github.io/warp/modules/contribution_guide.html

Please fill the relevant sections.

Checkboxes can also be marked after you submit the PR.
-->

## Description
<!--
Please add a description of what this PR aims to accomplish. 
Existing issues may be reference using a special keyword, e.g. Closes #10
Include any limitations or non-handled areas in the changes.
-->

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] Necessary tests have been added
- [x] Documentation is up-to-date
- [x] Auto-generated files modified by compiling Warp and building the documentation have been updated (e.g. `__init__.pyi`, `functions.rst`)
- [x] Code passes formatting and linting checks with `pre-commit run -a`

shi-eric commented on warp/tests/test_atomic_bitwise_tile.py:
Can we move this tile-specific test into warp/tests/tile?

daedalus5 commented:
Are these atomic ops needed for all vector / matrix types? If so, we'll have to add tests for these.
If they aren't needed, it might be better to remove them from this PR. We could still add at a later date if needed.

daedalus5 reviewed (commented):
Very comprehensive!

j3soon commented:
While adding the tests for vector and matrix types (for both global arrays and tiles), I found that my previous changes are incomplete. The following are the main changes since last review:

1. Added bitwise operators (bit_and/bit_or/bit_xor/lshift/rshift/invert) for vector and matrix types (following those supported by numpy), which are necessary for writing the tests. The correctness of most of the newly introduced operators are also validated in the testcases.
2. Added bitwise operators for inter-tiles operations such as `tile & tile` for tests.
4. Removed in-place bitwise operators for quaternion and spatial transform, as bitwise operations do not make sense on these data types.
3. Renamed functions such as `and(...)` to `bit_and(...)` and `and_inplace(...)` to `bit_and_inplace(...)`, so as to follow the convention of the codebase. Atomics are kept the original `atomic_and` following CUDA C++.

j3soon commented:
Thanks for reviewing! I've pushed several commits to resolve the comments. Please let me know if there are further issues.

shi-eric commented:
Since you opened this pull request on GitHub and we do merges on GitLab, we cannot squash your changes and merge because this results in you not getting credit for the work when looking at the Git log. It would help if you cleaned up your history using Git rebase so it's down to a single commit.

j3soon commented:
Got it! Will squash the commits and perform a rebase.

shi-eric commented:
Thank you @j3soon, I'll open a merge request with the as-is changes on our internal GitLab repo.

j3soon commented:
@shi-eric I've just squashed the changes into a single commit, and then rebase onto main. This PR (or commit) should be free of merge conflicts now! Sorry for the hassle, I'll learn the GitLab process next time. 🙏

daedalus5 commented:
Looks good @j3soon !

shi-eric commented on CHANGELOG.md:
```diff
@@ -38,6 +38,8 @@
 - Deprecate support for Intel-based macOS (x86_64) with removal targeted in late 2025.
   We will continue to support Apple Silicon-based Macs with the CPU backend.
   Users on Intel-based hardware will now receive a `DeprecationWarning` with this information.
+- Add `wp.atomic_and()` (`&=`), `wp.atomic_or()` (`|=`), and `wp.atomic_xor()` (`^=`) built-ins for atomic bitwise operations, along with bitwise operations for vector, matrix, and tile types.
+  ([GH-886](https://github.com/NVIDIA/warp/issues/886)).
```
We'll fix it ourselves.

j3soon commented on warp/native/mat.h:
```diff
@@ -2276,22 +2258,816 @@ inline CUDA_CALLABLE void adj_assign_copy(
+}
+
+
+template<unsigned Rows, unsigned Cols, typename Type>
```
Thanks to you both @shi-eric and @mmacklin for catching this! I just pushed a fix for this issue.

Specifically, I accidentally included a deleted matrix `mod` broadcasting operation during a rebase onto commit 0029c21f527d33abf1c60d5517a74cbf1aad0193.

https://github.com/NVIDIA/warp/blob/00d344d99d85f0f3664ef2913846c6d3a91331a7/warp/native/mat.h#L3698-L3711

where the above matrix `mod` function is deleted in this diff: https://github.com/NVIDIA/warp/commit/0029c21f527d33abf1c60d5517a74cbf1aad0193 but incorrectly re-added during a rebase without conflict. (I just learned that a non-conflicting rebase may still cause issues)

This is not caught before since it's an extra function not exposed anywhere else, and does not affect the correctness of the bitwise operation testcases. The latest commit removes this `mod` function.

As a side note, I'm not sure why the pull request diff of `mat.h` looks so broken on both GitHub/GitLab webpage diff. The VSCode diff seems much better here: https://vscode.dev/github/NVIDIA/warp/pull/887

shi-eric commented:
> Hi @shi-eric, I included the internal changes, rebased onto main, and squashed the commits.
> 
> Better diff (compared to GitHub/GitLab) here: https://vscode.dev/github/NVIDIA/warp/pull/887
> 
> Thanks!

Thanks! The link you shared is very useful indeed.
